Signal/src/ViewControllers/ConversationView/ConversationViewModel.m

```diff
@@ -0,0 +1,1284 @@
+    BOOL isOnlyModifyingLastMessage = YES;
+    for (ConversationUpdateItem *updateItem in updateItems) {
+        switch (updateItem.updateItemType) {
+            case ConversationUpdateItemType_Delete:
+                isOnlyModifyingLastMessage = NO;
+                break;
+            case ConversationUpdateItemType_Insert: {
+                id<ConversationViewItem> viewItem = updateItem.viewItem;
+                OWSAssertDebug(viewItem);
+                if (([viewItem.interaction isKindOfClass:[TSIncomingMessage class]] ||
+                        [viewItem.interaction isKindOfClass:[TSOutgoingMessage class]])
+                    && updateItem.newIndex >= oldViewItemCount) {
+                    continue;
+                }
+                isOnlyModifyingLastMessage = NO;
+                break;
+            }
+            case ConversationUpdateItemType_Update: {
+                id<ConversationViewItem> viewItem = updateItem.viewItem;
+                if ([updatedNeighborItemSet containsObject:viewItem.itemId]) {
+                    continue;
+                }
+                OWSAssertDebug(viewItem);
+                if (([viewItem.interaction isKindOfClass:[TSIncomingMessage class]] ||
+                        [viewItem.interaction isKindOfClass:[TSOutgoingMessage class]])
+                    && updateItem.newIndex >= oldViewItemCount) {
+                    continue;
+                }
+                isOnlyModifyingLastMessage = NO;
+                break;
+            }
+        }
+    }
+    BOOL shouldAnimateRowUpdates = !isOnlyModifyingLastMessage;
+    return shouldAnimateRowUpdates;
+}
+
+- (void)createNewMessageMappings
+{
+    if (self.thread.uniqueId.length > 0) {
+        self.messageMappings = [[YapDatabaseViewMappings alloc] initWithGroups:@[ self.thread.uniqueId ]
+                                                                          view:TSMessageDatabaseViewExtensionName];
+    } else {
+        OWSFailDebug(@"uniqueId unexpectedly empty for thread: %@", self.thread);
+        self.messageMappings =
+            [[YapDatabaseViewMappings alloc] initWithGroups:@[] view:TSMessageDatabaseViewExtensionName];
+    }
+
+    // Cells' appearance can depend on adjacent cells in both directions.
+    //
+    // TODO: We could do the same thing in our logic to generate "update items"
+    //       in updateViewWitholdItemIdList.
+    [self.messageMappings setCellDrawingDependencyOffsets:[NSSet setWithArray:@[
+        @(-1),
+        @(+1),
+    ]]
+                                                 forGroup:self.thread.uniqueId];
+
+    [self.uiDatabaseConnection readWithBlock:^(YapDatabaseReadTransaction *transaction) {
+        [self.messageMappings updateWithTransaction:transaction];
+    }];
+    // We need to impose the range restrictions on the mappings immediately to avoid
+    // doing a great deal of unnecessary work and causing a perf hotspot.
+    [self updateMessageMappingRangeOptions];
```

Comment:
Should this happen *before* `updateWithTransaction` to avoid said "perf hotspots"?

Reply:
The mappings have to be updated before you update their range.  That was very counter-intuitive to me and I never found docs the clearly set out the correct order of these operations, but the the results are wrong if you do it the other way.  

We solved bugs around this a while back.